src/resolve/resolved-mdns.c

```diff
@@ -262,12 +256,19 @@ int manager_mdns_ipv4_fd(Manager *m) {
                 goto fail;
         }
 
+        /* try to bind without SO_REUSEADDR option to detect another mDNS responder */
         r = bind(m->mdns_ipv4_fd, &sa.sa, sizeof(sa.in));
         if (r < 0) {
                 r = -errno;
                 goto fail;
         }
 
+        r = setsockopt(m->mdns_ipv4_fd, SOL_SOCKET, SO_REUSEADDR, &one, sizeof(one));
```

Comment:
Hm, we set SO_REUSEADDR so that if we are quickly restarted, we can bind to the same port immediately. SO_REUSEADDR has an effect during `bind`, afaiu. Setting it *afterward* has no effect. I expect that with this change, if systemd-resolved crashes with some mdns TCP transfers outstanding, or if the remote crashes, and we try to restart systemd-resolved, it would be able to bind to the socket.

The reason why the bind from two processes can succeed is because they attach to different addresses. So instead of disabling SO_REUSEADDR we should figure out what address avahi is `bind`-ing to, and also bind to that address. Then we should fail if avahi is listening.